### ISSUE TITLE
Svelte Typescript Port

### ISSUE THREAD

**Issue by roblevintennis** [AgnosticUI/agnosticui#127]
We'd like to convert the Svelte implementation to use Typescript since it not being so is likely a blocker for many modern web developers at this point. Typings and IDE hints.

I suspect that the easiest approach to this might be to use the svelte-kit starters to kickstart a new sub-directory package with typescript preconfigured; and then copy the components over one at a time.

Related to [Vue Typescript Port](https://github.com/AgnosticUI/agnosticui/issues/20)

**Comment by gcsecsey** [comment]
Hi! 👋 I'd love to tackle this. Similarly to #20, would it be OK here to temporarily have an extra in-progress package, and to incrementally add components in smaller PRs?

**Comment by roblevintennis** [comment]
@gcsecsey thanks for your interest here!

Yes, I think having an interim agnostic-svelte2 or something would make sense and be fine. We of course wouldn't publish it!

Once component and docs parity we'd just swap it in.

I'm also happy to help per component and divy these up if it's helpful.

Lmk if you have any questions

**Comment by gcsecsey** [comment]
> Yes, I think having an interim agnostic-svelte2 or something would make sense and be fine. We of course wouldn't publish it until we the component and docs parity at which point we'd just swap it in place of agnostic-svelte, bump the package version and publish.

This sounds like a great plan! 🎉  I'll start working on the interim solution on a branch and only raise a PR to upstream once we reach parity.

> I'm also happy to help per component and divy these up if it's helpful (lmk).

Thanks, I'll definitely reach out with questions once I get there! 🙂  

> I'm pretty sure the easiest way would be to just initialize a new svelte-kit project and select typescript and then port the components over one at a time. I've done this several times and it's pretty straightforward but a bit tedious 😄

Yes, I've started out just like that, and initialized storybook on the project. I'll publish these changes as soon as I get the chance, just run out of time yesterday. 😄

**Comment by roblevintennis** [comment]
_Yes, I think having an interim agnostic-svelte2 or something would make sense and be fine._

> This sounds like a great plan! 🎉 I'll start working on the interim solution on a branch and only raise a PR to upstream once we reach parity.

Sorry, I don't think I expressed this very clearly :-) Let me try again…

I'm actually thinking the very first PR should be:
1. `agnostic-svelte-ts/` with main svelte-kit boilerplates and perhaps Storybook 
2. Just one component. Preferably one of the simpler ones like Close

```shell
├── agnostic-svelte
└── agnostic-svelte-ts
    └── src
        └── lib
            └── components
                └── Close
                    ├── Close.stories.ts
                    └── Close.svelte
```

So once that was merged, yes, we'd have the extra package. When I said we wouldn't publish I meant `npm publish` 😄 

From there, we could do additional PR's for a few components at a time. This will make it much much easier to review.

Then, once we've reached parity, I can completely remove `agnostic-svelte` (old version) and replaces e.g.`rm -rf ./agnostic-svelte && mv agnostic-svelte-ts agnostic-svelte`.

That last step is something I think I'd prefer to do as I'll likely do one more verification step and `npm publish` update docs etc.

Another reason besides avoiding a huge PR is that we'd pretty much have to halt other fixes as I doubt it's practical to continually merge in bug fixes and feature updates (e.g. this is a pretty critical [issues](https://github.com/AgnosticUI/agnosticui/issues/147) which pretty much makes AgnosticUI almost useless for forms at the moment [my fault of course!] but I'd like to fix)

**Comment by gcsecsey** [comment]
> It sounds like you may be very close to having this already no?

Yes, I have the working copy of Sveltekit + TS set up, and storybook initialized, which still throws some errors, so I'm working on fixing the configs and dependency versions. For the time being, I've raised #151 which only has the Sveltekit set up, but I'm planning to add the storybook config and a component to it in later commits. 

> Does all above ☝️ I've laid out sound good to you? Lmk and thanks! 🙏

Yes, I think this is a good plan! 🎉 Thanks for pointing out the priorities as well, I'll avoid Inputs and Select for now and wait for the fix on these.

**Comment by gcsecsey** [comment]
Another question:
I've had some issues with committing the changes, as the pre-commit hooks runs eslint and prettier on all subprojects, I've run `cd agnostic-[subproject] && npm i`on all subprojects before commit. Still, I've received some errors:

In `agnostic-svelte`:
```
> agnostic-svelte@1.1.16 format
> npm run prettier --write './src/**/*.{js,svelte,md,html,json}'

npm ERR! Missing script: "prettier"
npm ERR! 
npm ERR! Did you mean this?
npm ERR!     npm run preview # run the "preview" package script
npm ERR! 
npm ERR! To see a list of scripts, run:
npm ERR!   npm run
```
 
In `agnostic-vue`:
```
> agnostic-vue@1.0.24 lint
> eslint --ext .js,.vue --ignore-path .gitignore src


Oops! Something went wrong! :(

ESLint: 8.12.0

Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/gcsecsey/projects/agnosticui/agnostic-vue/.eslintrc.js from /Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/@eslint/eslintrc/dist/eslintrc.cjs not supported.
.eslintrc.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead rename .eslintrc.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /Users/gcsecsey/projects/agnosticui/agnostic-vue/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).

    at Object.module.exports [as default] (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/import-fresh/index.js:32:59)
    at loadJSConfigFile (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:2559:47)
    at loadConfigFile (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:2643:20)
    at ConfigArrayFactory.loadInDirectory (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:2840:34)
    at CascadingConfigArrayFactory._loadConfigInAncestors (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3833:46)
    at CascadingConfigArrayFactory._loadConfigInAncestors (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3852:20)
    at CascadingConfigArrayFactory.getConfigArrayForFile (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3754:18)
    at FileEnumerator._iterateFilesRecursive (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/eslint/lib/cli-engine/file-enumerator.js:446:49)
    at _iterateFilesRecursive.next (<anonymous>)
    at FileEnumerator.iterateFiles (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/eslint/lib/cli-engine/file-enumerator.js:297:49)
    at iterateFiles.next (<anonymous>)
    at CLIEngine.executeOnFiles (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/eslint/lib/cli-engine/cli-engine.js:787:48)
    at ESLint.lintFiles (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/eslint/lib/eslint/eslint.js:565:23)
    at Object.execute (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/eslint/lib/cli.js:301:36)
    at main (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/eslint/bin/eslint.js:138:52)
    at Object.<anonymous> (/Users/gcsecsey/projects/agnosticui/agnostic-vue/node_modules/eslint/bin/eslint.js:142:2)
```

Is there something missing from my environment? Should I raise these as issues and create PRs for them, perhaps?

**Comment by roblevintennis** [comment]
@gcsecsey it sounds like you've found another couple bugs! Yes, can you log another issue and if you also have spare cycles to fix that's great 🙏 I think I may have removed some stuff and then not made adjustments to those hooks.

For here, if you'd like to just use `commit --no-verify -m "commit messages"` that's fine. I'll go take a look at the other PR now

**Comment by roblevintennis** [comment]
@gcsecsey I'm not sure if you were planning to take on the precommit bug that's preventing commits (less the workaround hack I mentioned), but this has been mentioned again today and so I'd say it's critical. I'm fine fixing if you'd prefer to stay focused. But I created https://github.com/AgnosticUI/agnosticui/issues/154 to track it so just lmk so we don't duplicate efforts 😉

**Comment by roblevintennis** [comment]
I'm not at my computer but I'll likely just remove the husky directory or whatever is kicking in the precommit (until things stabilize and we can reintroduce later...much later lol 😂)

**Comment by gcsecsey** [comment]
@roblevintennis I'm happy to try to tackle the precommit issue as well. I've added some updates on the issue. 🙂

**Comment by roblevintennis** [comment]
@gcsecsey yeah, I saw you just removed the husky directory which is what I was planning to do :) If you'd like to just cut a small PR I'll merge that.

## My excuse

I've been moving pretty quickly and done some huge overhauls a couple of months ago where I did things like completely switch from Vue 2 to Vue 3 + Vitepress, upgrade to Angular 13, upgrade to Svelte-Kit, etc., etc. I became extremely tedious to port all the tests and a lot of them were quite silly anyway, so I just nuked a lot of them.

Unfortunately, this means we don't have much test coverage which is something longer term to improve. However, these are fairly simple primitive ui components so I'm not entirely concerned about that.

Which brings as to why I think just nuking the husky is the right thing to do here. As things stabilize we can start to reintroduce.

Yes, I'd love a PR for that! Thanks 🙏

**Comment by roblevintennis** [comment]
@gcsecsey Heya, just checking back in to confirm you're still wanting to tackle this lmk

**Comment by roblevintennis** [comment]
Update: I think I'd like to see us break these out into individual component ports or a few related at a time so we can keep things manageable and allow for multiple folks working in tandem. For example buttons / button groups / buttons with icons could be one task, menues another, etc.

**Comment by gcsecsey** [comment]
@roblevintennis sorry I've been at a meetup last week and didn't have time to check on threads. 🌴😄

> Update: I think I'd like to see us break these out into individual component ports or a few related at a time so we can keep things manageable and allow for multiple folks working in tandem. For example buttons / button groups / buttons with icons could be one task, menues another, etc.

I like the direction of breaking this out to multiple smaller ports, sounds like a great plan both in terms of PR size and teamwork! 👍 

> Also, @Croug has mentioned he's interested in helping out in general so if you want to coordinate I think all of us could slowly pick off components. If we do want to go that route I would decompose stuff to the component level so we could ensure we're not duplicating efforts.

I'd like to proceed with adding storybook to the svelte installation and bringing the TS subproject to parity with the existing one so that we can start adding components. I'm expecting to be able to raise a PR with these changes this week, and will share any updates. 🙂

**Comment by roblevintennis** [comment]
Sounds good @gcsecsey 👌🏽 go for it! We've got plenty of other tasks to pull from 👍🏽

**Comment by Croug** [comment]
I think if we're going to go forward with splintering this into multiple component groups we should coordinate over on [gitter](https://gitter.im/AgnosticUI/contribute) so we can work without stepping on each others toes!

**Comment by roblevintennis** [comment]
I've pitched in a bit -- I added `svelte2tsx` and a few components. Will keep trying to pitch in when I have spare cycles

**Comment by roblevintennis** [comment]
@Croug @gcsecsey guys, given my absence I'm assuming y'all have likely "moved on" (no worries haha). I'm going to probably attempt finishing this (I believe it mainly means getting the Table + Pagination stuff ported over) but thought I'd at mention y'all just in case there was interest in being involved. Again, I totally understand if not at this point 😄
